### PR TITLE
Resource allocation

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Partial implementation of resource allocation module

Technical task : https://github.com/smartdevicelink/sdl_core/issues/1667
Requirements : https://github.com/smartdevicelink/sdl_requirements/blob/master/detailed_docs/resource_allocation.md

### CODE REVIEW

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ public:
+  /**
+   * @brief AquireResource aquires resource by applicaion
+   * @param module_type resource to aquire
+   * @param app_id applicaiton that aquire resource
+   * @return ALLOWED if resource aquired \
+   *         IN_USE  if subscription is not allowed
+   *         ASK_DRIVER is driver confirm is required
+   */
+  virtual AquireResult::eType AquireResource(const std::string& module_type,
```
Comment (LitvinenkoIra) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan Acquire 
fix it everywhere

File: src/components/remote_control/include/remote_control/remote_control_plugin.h [hunk 2]
```diff
@@ -101,32 +104,35 @@ class RemoteControlPlugin : public RemotePluginInterface {
...
+
   functional_modules::PluginInfo plugin_info_;
   bool is_scan_started_;
   request_controller::RequestController request_controller_;
-  rc_event_engine::EventDispatcher<application_manager::MessagePtr, std::string>
-      event_dispatcher_;
 
+  RCEventDispatcher event_dispatcher_;
+
+  ResourceAlocationManagerImpl resource_alocation_manager_;
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/remote_control_plugin.h]: @LuxoftAKutsan A*ll*ocation

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                                             uint32_t app_id) = 0;
+
+  /**
+   * @brief AskDriver send OnInteriorConcent request to HMI for aqquiring
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan consent

Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan acquiring. May be change to `to acquire resource`?

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    LOG4CXX_WARN(logger_, "App with app_id " << app_id << "does not exist!");
+    return AquireResult::IN_USE;
+  }
+
+  const AllocatedResources::const_iterator allocated_it =
+      allocated_resources_.find(module_type);
+  if (allocated_resources_.end() == allocated_it) {
+    allocated_resources_[module_type] = app_id;
+    LOG4CXX_DEBUG(logger_,
+                  "Resource is not ys aquired. "
```
Comment (AByzhynar) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: @LuxoftAKutsan `ys` ?

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                  "Allow "
+                      << app_id << " aquiring " << module_type);
+    return AquireResult::ALLOWED;
+  }
+
+  if (AccessMode::AUTO_DENY == current_access_mode_) {
```
Comment (AByzhynar) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: @LuxoftAKutsan May be move this check up?

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                  "Disallow "
+                      << app_id << " aquiring " << module_type);
+    return AquireResult::IN_USE;
+  }
+
+  const mobile_apis::HMILevel::eType aquiring_app_hmi_level =
+      aquiring_app->hmi_level();
+  if (mobile_apis::HMILevel::HMI_FULL != aquiring_app_hmi_level) {
+    LOG4CXX_DEBUG(logger_,
+                  "Aquiring resources now alllowed in "
```
Comment (AByzhynar) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: @LuxoftAKutsan typo

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                      << aquiring_app_hmi_level << " hmi level. "
+                      << "Disallow " << app_id << " aquiring " << module_type);
+    return AquireResult::IN_USE;
+  }
+
+  if (AccessMode::AUTO_ALLOW == current_access_mode_) {
```
Comment (AByzhynar) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: @LuxoftAKutsan We can use switch instead of ifs

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/remote_control/include/remote_control/remote_control_plugin.h (hunk 2 above)
Comment (LuxoftAKutsan) [reply src/components/remote_control/include/remote_control/remote_control_plugin.h]: fixed

File: src/components/remote_control/src/resource_allocation_manager_impl.cc (hunk 5 above)
Comment (LuxoftAKutsan) [reply src/components/remote_control/src/resource_allocation_manager_impl.cc]: @AByzhynar if policy id AUTODENY but resource is not acquired , SDL should allow acquire this resource

File: src/components/remote_control/src/resource_allocation_manager_impl.cc (hunk 7 above)
Comment (LuxoftAKutsan) [reply src/components/remote_control/src/resource_allocation_manager_impl.cc]: fixed

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  /**
+   * @brief AcquireResource forses aquiring resource by applicaion
+   * @param module_type resource to acquire
+   * @param app_id applicaiton that acquire resource
+   */
+  virtual void ForceAcquireResource(const std::string& module_type,
+                                    uint32_t app_id) = 0;
+
+  /**
+   * @brief OnDriverDisallowed callback for rejecting aquiring resource
```
Comment (LitvinenkoIra) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: acquiring

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+   * @param module_type resource type
+   * @param app_id applicaiton id
+   */
+  virtual void OnDriverDisallowed(const std::string& module_type,
+                                  uint32_t app_id) = 0;
+  /**
+   * @brief AskDriver send OnInteriorConcent request to HMI for aqquiring
```
Comment (LitvinenkoIra) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: acquiring

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+ResourceAllocationManagerImpl::ResourceAllocationManagerImpl(
+    RemotePluginInterface& rc_plugin)
+    : current_access_mode_(AccessMode::AUTO_ALLOW)
+    , active_call_back_(NULL)
+    , rc_plugin_(rc_plugin) {}
+
+AcquireResult::eType ResourceAllocationManagerImpl::AcquireResource(
+    const std::string& module_type, uint32_t app_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const application_manager::ApplicationSharedPtr aquiring_app =
```
Comment (LitvinenkoIra) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: acquiring_app

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  if (allocated_resources_.end() == allocated_it) {
+    allocated_resources_[module_type] = app_id;
+    LOG4CXX_DEBUG(logger_,
+                  "Resource is not acquired yet. "
+                  "Allow "
+                      << app_id << " aquiring " << module_type);
+    return AcquireResult::ALLOWED;
+  }
+
+  const mobile_apis::HMILevel::eType aquiring_app_hmi_level =
```
Comment (LitvinenkoIra) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: acquiring_app_hmi_level

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+  // execute callback on response
+  // event_dispatcher_.add_observer(function_id, msg[kId].asInt(), callback);
+  active_call_back_ = callback;
+}
+
+ResourceAllocationManagerImpl::~ResourceAllocationManagerImpl() {}
+
+void ResourceAllocationManagerImpl::ForceAcquireResource(
+    const std::string& module_type, uint32_t app_id) {
+  LOG4CXX_DEBUG(logger_, "Force " << app_id << " aquiring " << module_type);
```
Comment (LitvinenkoIra) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: acquiring

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+class ResourceAllocationManager {
+ public:
+  /**
+   * @brief AcquireResource acquires resource by applicaion
+   * @param module_type resource to acquire
+   * @param app_id applicaiton that acquire resource
+   * @return ALLOWED if resource acquired \
+   *         IN_USE  if subscription is not allowed
+   *         ASK_DRIVER is driver confirm is required
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan confirm -> confirmation or approval

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/remote_control/include/remote_control/resource_allocation_manager_impl.h [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  ResourceAllocationManagerImpl(RemotePluginInterface& rc_plugin);
+
+  AcquireResult::eType AcquireResource(const std::string& module_type,
+                                       uint32_t app_id) OVERRIDE FINAL;
+  void AskDriver(const std::string& module_type,
+                 uint32_t app_id,
+                 AskDriverCallBackPtr callback) OVERRIDE FINAL;
+
+  void SetAccessMode(AccessMode::eType access_mode);
+  ~ResourceAllocationManagerImpl() OVERRIDE FINAL;
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager_impl.h]: @LuxoftAKutsan `FINAL` && `OVERRIDE` are redundant for `dtor`

File: src/components/remote_control/src/commands/base_command_request.cc [hunk 15]
```diff
@@ -94,13 +95,95 @@ void BaseCommandRequest::SendResponse(bool success,
   }
 }
 
+struct OnDrivesAnswerImpl : AskDriverCallBack {
```
Comment (AByzhynar) [inline src/components/remote_control/src/commands/base_command_request.cc]: @LuxoftAKutsan OnDrives???

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/remote_control/src/commands/base_command_request.cc [hunk 16]
```diff
@@ -94,13 +95,95 @@ void BaseCommandRequest::SendResponse(bool success,
...
+        service_.ValidateMessageBySchema(hmi_response);
+    LOG4CXX_DEBUG(logger_,
+                  "HMI response validation result is " << validate_result);
+    if (validate_result !=
+        application_manager::MessageValidationResult::SUCCESS) {
+      return;
+    }
+    const Json::Value value =
+        MessageHelper::StringToValue(hmi_response.json_message());
+    bool allowed = value[message_params::kAllowed].asBool();
```
Comment (AByzhynar) [inline src/components/remote_control/src/commands/base_command_request.cc]: @LuxoftAKutsan const?

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/remote_control/include/remote_control/commands/set_interior_vehicle_data_request.h [hunk 17]
```diff
@@ -60,6 +60,9 @@ class SetInteriorVehicleDataRequest : public BaseCommandRequest {
    */
   void Execute() FINAL;
 
+  AcquireResult::eType AcquireResource(
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/commands/set_interior_vehicle_data_request.h]: @LuxoftAKutsan Add description please

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+      LOG4CXX_DEBUG(logger_,
+                    "Current access_mode is AUTO_DENY. "
+                    "Disallow "
+                        << app_id << " aquiring " << module_type);
+      return AcquireResult::IN_USE;
+    }
+    case AccessMode::ASK_DRIVER: {
+      LOG4CXX_DEBUG(logger_,
+                    "Current access_mode is ASK_DRIVER. "
+                    "Driver confirm required for allowing "
```
Comment (AByzhynar) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: @LuxoftAKutsan Driver confirm required for allowing " - > Driver confirmation required to allow "

File: src/components/remote_control/src/resource_allocation_manager_impl.cc [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+                        << app_id << " aquiring " << module_type);
```
Comment (AByzhynar) [inline src/components/remote_control/src/resource_allocation_manager_impl.cc]: @LuxoftAKutsan to acquire

File: src/components/remote_control/include/remote_control/commands/base_command_request.h [hunk 20]
```diff
@@ -91,6 +91,14 @@ class BaseCommandRequest
   application_manager::MessagePtr message_;
   Json::Value response_params_;
 
+  /**
+   * @brief AcquireResource try to allocate resource for application
+   * In case if allocation of resource is not required, return ALLOWD by
```
Comment (LitvinenkoIra) [inline src/components/remote_control/include/remote_control/commands/base_command_request.h]: ALLOWED

**Review by LitvinenkoIra — APPROVED** [review]

File: src/components/remote_control/include/remote_control/remote_plugin_interface.h [hunk 21]
```diff
@@ -116,6 +118,8 @@ class RemotePluginInterface : public functional_modules::GenericModule {
 
   virtual RCPluginEventDispatcher& event_dispatcher() = 0;
 
+  virtual ResourceAllocationManager& resource_allocator_manager() = 0;
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/remote_plugin_interface.h:121]: @LuxoftAKutsan resource_allocator_manager()-> resource_allocation_manager()

**Review by AByzhynar — CHANGES_REQUESTED** [review]

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Enum for list of results of allocation resources
+ */
+namespace AcquireResult {
+enum eType { ALLOWED = 0, IN_USE, ASK_DRIVER };
+}
+
+/**
+ * @brief The AskDriverCallBack class callback for GetInteriourConcent response
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan concent ->consent

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+ public:
+  virtual ~AskDriverCallBack() {}
+};
+
+typedef utils::SharedPtr<AskDriverCallBack> AskDriverCallBackPtr;
+
+class ResourceAllocationManager {
+ public:
+  /**
+   * @brief AcquireResource acquires resource by applicaion
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan application insted of applicaion

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+   * @param module_type resource to acquire
+   * @param app_id applicaiton that acquire resource
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan application insted of applicaion

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+   * @return ALLOWED if resource acquired \
+   *         IN_USE  if subscription is not allowed
+   *         ASK_DRIVER if driver confirmation is required
+   */
+  virtual AcquireResult::eType AcquireResource(const std::string& module_type,
+                                               uint32_t app_id) = 0;
+
+  /**
+   * @brief AcquireResource forses acquiring resource by applicaion
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan application insted of applicaion

Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan forCes

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+   * @param module_type resource to acquire
+   * @param app_id applicaiton that acquire resource
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan application insted of applicaiton. Check everywhere, please

File: src/components/remote_control/include/remote_control/resource_allocation_manager.h [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+   * @param app_id applicaiton id
+   */
+  virtual void OnDriverDisallowed(const std::string& module_type,
+                                  uint32_t app_id) = 0;
+  /**
+   * @brief AskDriver send GetInteriorConcent request to HMI for acquiring
+   * resource
+   * @param module_type resource to acquire
+   * @param app_id applicaiton that acquire resource
+   * @param callback will be executed on OnInteriorConcent response
```
Comment (AByzhynar) [inline src/components/remote_control/include/remote_control/resource_allocation_manager.h]: @LuxoftAKutsan conSent

**Review by AByzhynar — APPROVED** [review]

**Pull request merged**